dali/base/dasds.cpp

```diff
@@ -1143,251 +1261,212 @@ class CBackupHandler : public CInterface, implements IThreaded
-    {
-        backupPath.set(_backupPath);
-        async = _async;
+        iStoreHelper = _iStoreHelper;
+        iStoreHelper->getPrimaryLocation(dataPath);
+        iStoreHelper->getBackupLocation(backupPath);
+
+        saveThresholdSecs = config.getPropInt("@deltaSaveThresholdSecs", defaultSaveThresholdSecs);
+        transactionQueueLimit = config.getPropInt("@deltaTransactionQueueLimit", defaultDeltaTransactionQueueLimit);
+        unsigned deltaTransactionMaxMemMB = config.getPropInt("@deltaTransactionMaxMemMB", defaultDeltaMemMaxMB);
+        transactionMaxMem = (memsize_t)deltaTransactionMaxMemMB * 0x100000;
+        if (saveThresholdSecs)
+            thresholdDuration = queryOneSecCycles() * saveThresholdSecs;
+        if (0 == transactionQueueLimit) // treat 0 as disabled, meaning that the # pending transactions will not be considered
+            transactionQueueLimit = INFINITE;
+
         aborted = false;
-        PROGLOG("BackupHandler started, async=%s", async?"true":"false");
-        threaded.init(this);
-    }
-    void stop()
-    {
-        if (!aborted)
-        {
-            aborted = true;
-            pending.signal();
-            threaded.join();
-        }
+        waiting = true;
+        VStringBuffer msg("CDeltaWriter started - deltaSaveThresholdSecs=%u, deltaTransactionMaxMemMB=%u", saveThresholdSecs, deltaTransactionMaxMemMB);
+        msg.append(", deltaTransactionQueueLimit=");
+        if (INFINITE == transactionQueueLimit)
+            msg.append("<DISABLED>");
+        else
+            msg.append(transactionQueueLimit);
+        PROGLOG("%s", msg.str());
+
+        if ((transactionQueueLimit > 1) && (transactionMaxMem > 0))
+            threaded.init(this);
+        else
+            PROGLOG("All transactions will be committed synchronously");
     }
-    void removeExt(const char *fname)
+    void addDelta(char *path, IPropertyTree *delta)
     {
-        if (aborted) return;
-        if (!async)
-        {
-            deleteExt(fname);
-            return;
-        }
-        BackupQueueItem *item = getFreeItem();
-        item->text->append(fname);
-        item->flags = BackupQueueItem::f_delext;
-        add(item);
+        CTransactionItem *item = new CTransactionItem(path, delta);
+        CriticalBlock b(pendingCrit);
+        addToQueue(item);
     }
-    void addExt(const char *fname, unsigned length, void *data)
+    CTransactionItem *addExt(char *name, size32_t length, void *content)
     {
-        if (aborted) return;
-        if (!async)
-        {
-            writeExt(fname, length, data);
-            free(data);
-            return;
-        }
-        BackupQueueItem *item = getFreeItem();
-        item->text->append(fname);
-        item->dataLength = length;
-        item->data = data; // take ownership
-        item->flags = BackupQueueItem::f_addext;
-        add(item);
+        CTransactionItem *item = new CTransactionItem(name, length, content);
```

Comment:
Should this be checking aborted still?  (And the other similar functions)

Reply:
I think it highlights an issue. The deltaWriter is cleared when blockingSave is called (when Dali itself saves all in memory state, the pending 'to be saved' transactions are cleared). However, when Dali shuts down that is not done.
The delta writer should both be cleared and stopped, although by the time it (Dali) is stopping and saving, there will be no more transactions added.

I will make some changes.